poly/src/polynomial/univariate/dense.rs

```diff
@@ -39,18 +41,63 @@ impl<F: Field> Polynomial<F> for DensePolynomial<F> {
     fn evaluate(&self, point: &F) -> F {
         if self.is_zero() {
             return F::zero();
+        } else if point.is_zero() {
+            return self.coeffs[0];
         }
-        let mut powers_of_point = vec![F::one()];
-        let mut cur = *point;
-        for _ in 0..self.degree() {
-            powers_of_point.push(cur);
-            cur *= point;
+        self.internal_evaluate(point)
+    }
+}
+
+#[cfg(feature = "parallel")]
+// Set some minimum number of field elements to be worked on per thread
+// to avoid per-thread costs dominating parallel execution time.
+const MIN_ELEMENTS_PER_THREAD: usize = 16;
+
+impl<F: Field> DensePolynomial<F> {
+    #[cfg(not(feature = "parallel"))]
+    fn internal_evaluate(&self, point: &F) -> F {
+        // Horners method - serial method
+        let mut result = F::zero();
+        let num_coeffs = self.coeffs.len();
+        for i in (0..num_coeffs).rev() {
+            result *= point;
+            result += self.coeffs[i];
         }
-        assert_eq!(powers_of_point.len(), self.coeffs.len());
-        ark_std::cfg_into_iter!(powers_of_point)
-            .zip(&self.coeffs)
-            .map(|(power, coeff)| power * coeff)
-            .sum()
+        result
+    }
+
+    #[cfg(feature = "parallel")]
+    fn internal_evaluate(&self, point: &F) -> F {
+        // Horners method - parallel method
+        // compute the number of threads we will be using.
+        let num_cpus_available = rayon::current_num_threads();
+        let num_coeffs = self.coeffs.len();
+        let num_elem_per_thread = max(num_coeffs / num_cpus_available, MIN_ELEMENTS_PER_THREAD);
+        // compute num_cpus_used as ceil(num_coeffs / num_elem_per_thread)
+        let num_cpus_used = (num_coeffs + num_elem_per_thread - 1) / num_elem_per_thread;
+
+        // run Horners method on each thread as follows:
+        // 1) Split up the coefficients across each thread evenly.
+        // 2) Do polynomial evaluation via horner's method for the thread's coefficeints
+        // 3) Scale the result point^{thread coefficient start index}
+        // Then obtain the final polynomial evaluation by summing each threads result.
+        let mut per_thread_res = vec![F::zero(); num_cpus_used];
+        let result = per_thread_res
+            .par_iter_mut()
+            .enumerate()
+            .map(|(i, _)| {
```

Comment:
```suggestion
        let result = (0..num_cpus_used)
            .map(|i| {
```

Reply:
Also, maybe this is better written using `par_chunks`?

Reply:
Thats a good point, I'll refactor using `par_chunks`. (I didn't know it existed when I first wrote this)